src/QsCompiler/Transformations/CallGraph.cs

```diff
@@ -153,6 +153,18 @@ public CallGraphEdge(TypeParameterResolutions paramResolutions)
                 kvp => StripPositionInfo.Apply(kvp.Value));
         }
 
+        /// <summary>
+        /// Copy constructor for CallGraphEdge objects.
+        /// Throws ArgumentNullException if edge is null or its ParamResolutions field is null
+        /// </summary>
+        public CallGraphEdge(CallGraphEdge edge)
+        {
+            if (edge == null || edge.ParamResolutions == null) throw new ArgumentNullException(nameof(edge));
+
+            // Remove position info from type parameter resolutions
+            ParamResolutions = edge.ParamResolutions;
```

Comment:
>ParamResolutions = edge.ParamResolutions [](start = 12, length = 40)

The new CallGraphEdge and `edge` now share ParamResolutions object, right? Is that OK? Is ParamResolutions field ever modified by the owning CallGraphEdge? If not, and sharing of the parameter resolutions is expected to be common, would it make sense to add object equality comparison into the Equal before comparing as sets?

Reply:
The ParamResolutions is a readonly Immutable Dictionary on the CallGraphEdge class, so I believe it cannot be changed and sharing should be fine. I'm not sure I understand your comment about the object equality. Could you dive into more detail about that?

Reply:
I meant that if it's common for object to share ParamResolutions you could compare the objects for equality of pointers first before comparing them as sets.

---
In reply to: [439695463](https://github.com/microsoft/qsharp-compiler/pull/454#discussion_r439695463) [](ancestors = 439695463)

Reply:
Ah I see. That makes for a nice short-circuit. I've added this check to the equality comparison.